### ISSUE TITLE
Can't change ajax parameters on-the-fly

### ISSUE THREAD

**Issue by cdaringe** [pouchdb/pouchdb#5322]
### Environment (Node.js/browser/hybrid app/etc.)

node.
### Adapter (IndexedDB/WebSQL/LevelDB/etc.)

all.
### Server (CouchDB/Cloudant/Couchbase/PouchDB Server/etc.)

n/a.
### problem statement / request

i need to be able to wrap `pouchdb-ajax`.

currently I can modify the inputs to the request module via `new PouchDB(name, { ajax: { ... } })`.  this is a onetime, static customization of the request.

i need runtime, dynamic ability to change the headers.  our authentication scheme requires headers adjustment periodically on sub-sequent requests.  I used to be able to wrap `PouchDB.utils.ajax` to achieve this.  unfortunately, this capability is now removed.

can we consider restoring the passing of a function to `{ ajax: ... }`?  thanks team!  you guys rock

**Comment by daleharvey** [comment]
So it was likely that I removed this when refactoring the http adapter but surprised this is the first report about it, @yaronyg wasnt this something thali depended on?

I have seen a few use cases for this and in a step towards having the integration tests run purely against the supported API I think it would be good to reintroduce this in a way that supports the tests + users use cases: 

 1. A simple ability to count requests then pass them along as they were
 2. A way to dynamically modify requests, to add a header for example
 3. To return custom data and not make the request at all

**Comment by yaronyg** [comment]
I vaguely remember that we did use this about a billion years ago when we were running in a webview against couchbase. But now we use node so all our work is there and we don't use ajax directly anymore.

**Comment by nolanlawson** [comment]
This can currently be supported by overriding `db._ajax`. It's not a primary use case that we support, though, hence the leading underscore.

Another way you can accomplish this is by using libraries that override `require()` and then sub in their own implementations, e.g. [nock](https://github.com/node-nock/nock) does this. Yet another way is to override the global `XMLHttpRequest` or `fetch` variables in the browser.

Personally I would prefer to close this as a wontfix since I can't think of any easy way to support this, and I'm not sure what the value-add is.

**Comment by ssured** [comment]
I used to have an indicator in my app which showed the upload and download speed of PouchDB syncing data. This was implemented by overriding utils.ajax. It provided a nice extra feedback to the end user about sync state. People seem to like having a little insight into what is happening; the syncing of PouchDB feels like magic to them. Having these counters again would be nice, is it something to add to PouchDB itself?

**Comment by cdaringe** [comment]
I did not title this issue well.  The title presumes a solution instead of describing the problem.  The problem is that I need to be able to modify ajax requests.  clearly, the simple existence of the `options.ajax` field is an acknowledgement that "sure, web devs need to be able to control their webrequests."  The problem is that I can only modify an ajax request _once_, on initialization.  my use case is already described above, where our authentication scheme yield a JWT that expires.  continued use of the services refreshed the JWT, hence, I want to update it s.t. my pouches continue to sync, even after expiration.

@ssured's use case is a great--we do that too.  it's actaully a different problem, albeit solved through the same means.

so, no, I don't need to wrap `pouchdb-ajax`, but that's how I did it before.  another solution, which may be a lot easier, it so add a lifecycle hook [here perhaps?](https://github.com/pouchdb/pouchdb/blob/master/packages/pouchdb-ajax/src/ajaxCore.js#L88).

maybe `options.preReqeust` is a thing.  then there on L88 we can do a

```js

**Comment by nolanlawson** [comment]
Hm yeah, I see a few ways to solve this problem:

1. make `pouchdb-ajax` pluggable, allow users to supply their own (hard)
2. add a `db.setAjaxOptions()` API so you're not limited to constructor time
3. `options.preRequest` or similar pre-request hook

Personally I'm in favor of #2 since it's the simplest conceptually and would probably add ~3 lines of code (we'd just need to make [this](https://github.com/pouchdb/pouchdb/blob/6cc3d9201b709882a2207c6d75b53b4ca1fb8953/packages/pouchdb-adapter-http/src/index.js#L149) a member variable), although I dislike exposing a new API method. @daleharvey ?

**Comment by mkazlauskas** [comment]
In my case, using `Authorization` header, I have to re-authenticate the user, which is an async operation. I suggest an async (returning promise?) version of `options.preRequest` or similar.

**Comment by nolanlawson** [comment]
Hm, why not re-authenticate yourself, and then update the ajax options in your own callback?

**Comment by mkazlauskas** [comment]
That is an option, but re-authentication would be invoked on some kind of error event, and with `retry` option it would continue making "hopeless requests". Reauthentication is only needed for (re)initiating replication, not keeping it, and it won't make much sense to keep constant authentication with JWTs if it's not required by application (with pouchdb it isn't).

If there was an async hook prior to request, I could do authentication on-demand.

**Comment by daleharvey** [comment]
So yeh I believe the old version was as @mkazlauskas said, with 

```
var db = new PouchDB('name', {
  ajax: function(opts, callback) { 
  }
});
```

I think that should be how it is exposed because the adapter/db api is generally symetrical across the adapters, this is http specific, also its going to be a comparatively rarely used so keeping the core api smaller is nicer imo, however I think we should make 2 changes, one is that we use promises, because promises, I also think we should pass in our implementation of ajax for the use case that the caller wants to modify the request but still make it, so:

```
var requestCount = 0;
var db = new PouchDB('name', {
  ajax: function(opts, ajax) {
     requestCount++;
     return ajax(opts);
  }
});
```

or 

```
var db = new PouchDB('name', {
  ajax: function(opts, ajax) {
     return Promise.reject({
       status: 412, 
       name: 'myfakeerror'
     });
  }
});
```

or 

```
var db = new PouchDB('name', {
  ajax: function(opts, ajax) {
     return doMyAuth().then(function(res) { 
       opts.headers.auth = res.auth;
       return ajax(opts);
     });
  }
});
```

Will need a little refactor of how we currently handle ajax internally, but totally backwards compatible and I think fairly clean + useful?

**Comment by nolanlawson** [comment]
:+1: on promise-only and attaching it to a constructor; I suppose it's best to provide the utmost flexibility here.

**Comment by DanielMaly** [comment]
Are there any updates on this issue? I'm deciding whether to hack db._ajax or wait for an official resolution.

**Comment by yaronyg** [comment]
I know that at some point we are going to get around to fixing this but honestly it's not likely to be for a month or two. :(

**Comment by daleharvey** [comment]
Working on this now, so one issue is `opts.ajax` is already a thing, so this needs a new name, using `req` for now but could probably do better, another is that this actually deprecates the need for `opts.ajax` because it can be done via `req`

```js
var db = new PouchDB('name', {
  ajax: function(opts, ajax) {
     opts.headers.myauth = myAuthToken;
     return ajax(opts);
  }
});
```

However removing `opts.ajax` or replacing it with this functionality would be a breaking change, so I think we should add this new functionality with a different name and deprecate `opts.ajax` and we can take it out in a few releases whenever we do a new purge release

**Comment by nolanlawson** [comment]
@daleharvey let's avoid breaking changes if we can; otherwise it's really hard to update the docs, every StackOverflow answer, etc. I'd be in favor of `opts.request` or `opts.intercept` or basically anything else. Alternatively we can detect `ajax: Object` vs `ajax: Function` and just automatically do the right thing in either case.

**Comment by nolanlawson** [comment]
currently blocked by #5835 because we want to make `fetch()` the target for hooks rather than `ajax()`/XHR

**Comment by gr2m** [comment]
Hey just wanted to let you know that in Hoodie, we resolved dynamic headers by using a getter: https://github.com/hoodiehq/hoodie-client/blob/a4e3544c889cd3145f37ab837adb70ad1ce04ac9/lib/get-api.js#L20-L33

**Comment by lgleim** [comment]
Has any of the proposed fixes been merged into Master by now? I would really like to have a proper way to set the Authorization header to do JWT bearer token authentication.

**Comment by gr2m** [comment]
@lgleim the solution linked in my comment above works very well. Any reason you can't use it for your case?

**Comment by lgleim** [comment]
Thanks for the pointer. It actually is a really nice monkey patch and does work well for the use-case. When I glanced at your link before I for some reason discarded it as a depending on a custom adapter implementation, i.e. Hoodie. ¯\_(ツ)_/¯

**Comment by gr2m** [comment]
it’s not a monkey patch, it does not change any internal behavior of PouchDB APIs, I’d say it’s a pretty clean solution :)

**Comment by cdaringe** [comment]
i agree, @gr2m's strategy is valid. closing

**Comment by tyler-johnson** [comment]
So the hack mentioned by @gr2m no longer works in PouchDB 6.1.2, making this a rather large issue for us.

It seems that PouchDB likes to deep clone the options object and this makes it impossible to hack. I have tried all sorts of things, from ES6 Proxies to defining classes to trick the cloner, but none seem to work.

The way I see it there are a few places that are blocking me from dynamic headers:

- [This line](https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-adapter-http/src/index.js#L150) deep clones the options object. Basically the only way to not have it deep clone is to create the headers or ajax object from a class, as [the cloner will avoid those](https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-utils/src/clone.js#L32). BUT
- [The `Object.assign` on these lines](https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-adapter-http/src/index.js#L167) is mixing the options *directly into the original*. This means that those objects *have to be cloned* otherwise the Object.assign will write to those object directly.

So whatever happen to the `request()` option? It really is the perfect solution for this issue.

**Comment by tyler-johnson** [comment]
Okay patching the `_ajax` method does work. For anyone stumbling up this now, this is how I did it:

```js
const db = new PouchDB("http://localhost:5984/mydb");

const ajax = db._ajax;
db._ajax = function(opts, cb) {
  const headers = Object.assign({}, opts.headers, myDynamicHeaders());
  return ajax(Object.assign({}, opts, { headers }), cb);
};
```

**Comment by daleharvey** [comment]
Reopening as this is something would still like to support officially

**Comment by daleharvey** [comment]
and yeh I swear I will get this one done

**Comment by daleharvey** [comment]
So this is happning in https://github.com/pouchdb/pouchdb/issues/6944, with a WIP PR posted that is pretty much working
